torch/_export/serde/serialize.py

```diff
@@ -633,15 +632,135 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
+        target_upgraders: List[Tuple[str]] = []
+        for ver in range(model_ver, curr_ver + 1):
+            if ver in versioned_upgraders:
+                target_upgraders.append(versioned_upgraders[ver])
+
+        return target_upgraders
+
+    def _populate_passes(self) -> List[UpgraderPass]:
+        """Given a list of upgraders, loop through it from lower version to higher version and create passes for all
+        upgraders. se torch.Library API to register old ops. Op name will be
+        <name>_<valid_from_ver>_<valid_till_ver>. Register upgarders as CompositeImplicitAutograd kernels. For example:
+
+        lib = Library("aten", "FRAGMENT")
+        lib.define(old_schema)
+
+        impl_lib = Library("aten", "IMPL")
+        impl_lib.impl("div__Scalar_0_3", div__Scalar_0_3, "CompositeImplicitAutograd")
+        """
+        lib = Library("aten", "FRAGMENT")
+        impl_lib = Library("aten", "IMPL")
+        upgrader_passes = []
+
+        def register_old_op(name: str, schema: str, impl_str: str):
+            """Registers an old version operator using impl_name as old op name."""
+            lib.define(schema)
+            exec(impl_str)
+            impl_lib.impl(name, locals()[name], "CompositeImplicitAutograd")
+
+        for schema, upgrader_str in self.upgraders:
+            upgrader_name = upgrader_str.split('(')[0].split(' ')[-1]
+            op_name = schema.split('(')[0].split("::")[-1]
+            schema = schema.replace(op_name, upgrader_name)
+            try:
+                register_old_op(name=upgrader_name, schema=schema, impl_str=upgrader_str)
+            except RuntimeError as e:
+                if "with the same name and overload name multiple times" in str(e):
+                    print(f"Registering {upgrader_name} multiple times")
+                else:
+                    raise RuntimeError from e
+            old_op_target = getattr(torch.ops.aten, upgrader_name).default
+            op_name, overload_name = (op_name, "default") if "." not in op_name else tuple(op_name.split(".")[:2])
+            new_op_target = getattr(getattr(torch.ops.aten, op_name), overload_name)
+            # Note that the graph will have op names in the graph, but actually they are of old versions.
+            upgrader_passes.append(GraphModuleOpUpgrader.UpgraderPass(old_target=new_op_target, new_target=old_op_target))
+
+        return upgrader_passes
+
+    def upgrade(self, exported_program: ep.ExportedProgram) -> ep.ExportedProgram:
+        args = [n.meta["val"] for n in exported_program.graph.nodes if n.op == "placeholder"]
+        inputs = tree_unflatten(args, exported_program.call_spec.in_spec)
+        from torch._dynamo import config
+        config.traceable_tensor_subclasses.add(FakeTensor)
+        for _pass in self.upgrader_passes:
+            exported_program = exported_program.transform(_pass).transform(partial(export, args=inputs))
```

Comment:
What's the second transform again? Export again? If so, shouldn't it happen once at the very end? Also missing other things passed to export, such as `constraints`?

Unless I completely misunderstood...

Reply:
Added comments, hopefully it's more clear.